relayer/relayer/src/listeners/eth.rs

```diff
@@ -61,125 +62,142 @@ impl EthereumListener {
-                                                return Ok(cb_event?);
-                                            },
-                                            ack_result = events_ack_receiver => {
-                                                if ack_result.is_ok () {
-                                                    info!(target: "EthereumListener", "Events ack received, marking {} as the next unprocessed block number", to_block +1);
-                                                    // we send + 1 to self as this is the next block we'd like to see
-                                                    last_processed_block_number.send(to_block + 1)?;
-                                                }
-                                            }
-                                        }
-                                    }
-                                }
-                            }
-                        }
-                    }
+            if events.is_empty() {
+                info!(target: LOG_TARGET, "Marking {} as the next unprocessed block number", to_block +1);
+                // we send + 1 to self as this is the next block we'd like to see
+                last_processed_block_number.send(to_block + 1)?;
+                continue;
+            }
+            let (events_ack_sender, events_ack_receiver) = oneshot::channel::<()>();
+            info!(target: LOG_TARGET, "Sending a batch of {} events", &events.len());
+
+            eth_events_sender
+                .send(EthMostEvents {
+                    events: events.clone(),
+                    from_block: unprocessed_block_number,
+                    to_block,
+                    events_ack_sender,
+                })
+                .await?;
+
+            info!(target: LOG_TARGET, "Awaiting events ack");
+
+            // select between ack and the channel, because the handler could have exited
+            select! {
+                cb_event = circuit_breaker_receiver.recv () => {
+                    warn!(target: LOG_TARGET, "Exiting before events ack due to a circuit breaker event {cb_event:?}");
+                    return Ok(cb_event?);
                 },
+                ack_result = events_ack_receiver => {
+                    if ack_result.is_ok () {
+                        info!(target: LOG_TARGET, "Events ack received, marking {} as the next unprocessed block number", to_block + 1);
+                        // we send + 1 to self as this is the next block we'd like to see
+                        last_processed_block_number.send(to_block + 1)?;
+                    }
+                }
             }
         }
     }
 }
 
-pub async fn get_next_finalized_block_number_eth(
+#[cfg(feature = "evm")]
+pub async fn get_next_finalized_block_number(
+    eth_connection: Arc<EthConnection>,
+    not_older_than: u32,
+) -> u32 {
+    // In evm context we treat latest block as finalized.
```

Comment:
Why it is so? Which stage are we (https://docs.arbitrum.io/how-arbitrum-works/tx-lifecycle) ?
To ask this question differently: are we fine with 'soft finality' here (https://docs.arbitrum.io/for-devs/troubleshooting-building#how-many-block-numbers-must-we-wait-for-in-arbitrum-before-we-can-confidently-state-that-the-transaction-has-reached-finality)

Reply:
Yep, we are fine with it :p more details in the channel dedicated to this topic